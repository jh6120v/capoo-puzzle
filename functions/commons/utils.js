
/**
 * 取一逆序列數和為 0 的一維陣列
 *
 * @param total
 * @returns {{label: number}[]}
 */
const getTiles = (total) => {
    return Array(total).fill(0).map((value, index) => ({ label: index, position: index }));
};

/**
 * 將數列隨機打散後，並檢查其是否可用(可解)
 *
 * @param tiles
 * @param cols
 * @returns {*}
 */
const getAcceptableTiles = (tiles, cols) => {
    let resolvable = false;

    while (!resolvable) {
        tiles.sort(() => {
            return Math.random() > 0.5 ? -1 : 1;
        });

        resolvable = checkResolvable(tiles, cols);
    }

    return tiles;
};

/**
 * 檢查數列是否有解，條件 e.g.
 * 1. 拼圖列數為奇數時，逆序列數和應為偶數
 * 2. 拼圖列數為偶數時，逆序列數和的奇偶性與空白磚所在列數的奇偶性守恆
 *
 * @param grids
 * @param cols
 * @returns {boolean}
 */
const checkResolvable = (grids, cols) => {
    // 逆序列數和
    let count = 0;

    // 空白磚所在列數
    let spaceX = 0;

    // 找出空白磚列數，並從 grids 中移除
    grids = grids.filter((item, idx) => {
        if (item.label === grids.length - 1) {
            spaceX = idx % cols + 1;
        }

        // 把空白從數列中去除(即數列中最後一數字)
        return item.label !== grids.length - 1;
    });

    // 計算逆序列數和
    grids.forEach((item, idx, grids) => {
        let j = idx;
        while (j < grids.length - 1) {
            if (item.label > grids[j + 1].label) {
                count++;
            }

            j++;
        }
    });

    return cols % 2 ? count % 2 === 0 : count % 2 + spaceX % 2 === 0;
};

/**
 *
 * @param cols
 * @returns {*}
 */
exports.getGrids = (cols) => {
    const tiles = getTiles(cols * cols);
    const grids = getAcceptableTiles(tiles, cols);

    return grids.reduce((data, item, idx) => {
        data.push({
            ...item,
            position: idx
        });

        return data;
    }, []);
};

const getPosition = (position, col) => {
    return {
        x: position % col,
        y: Math.floor(position / col)
    };
};

exports.getLayoutPositionList = (puzzleWidth, cols) => {
    const totalCols = cols * cols;
    const singleWidth = puzzleWidth / cols;

    return Array(totalCols).fill(0).map((value, index) => index).map(n => {
        const p = getPosition(n, cols);

        return { x: singleWidth * p.x, y: singleWidth * p.y };
    });
};

exports.LEVEL_MAP = {
    easy: 3,
    medium: 4,
    hard: 5
};
