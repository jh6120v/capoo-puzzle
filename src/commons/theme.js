export const theme = (darkMode) => {
    const lightTheme = {
        background: '#6cb8cf',
        textColor: '#1d1d1f',

        headerBg: '#6cb8cf',
        headerTextColor: '#ffffff',

        containerBg: '#aadff0',

        personalRecordItemTitleTextColor: '#4f6571',
        personalRecordItemContentBg: '#6cb8cf',
        personalRecordItemContentTextColor: '#ffffff',

        ratingTextColor: '#4f6571',

        settingBg: '#9fd4e5',
        settingItemBg: '#AADFF0',
        settingItemTextColor: '#4f6571',
        settingItemBorderColor: '#6cb8cf',
        settingItemIconColor: '#4f6571',
        settingItemInActiveBg: '#9FD1E0',

        versionTextColor: '#4f6571',

        modelBg: '#ffffff',
        modelTextColor: '#1d1d1f',
        modelConfirmBtnBg: '#5AC9E8',
        modelConfirmBtnTextColor: '#ffffff',

        functionBtnTextColor: '#4f6571',

        rankingBg: '#9fd4e5',
        rankingBorderColor: '#6cb8cf',
        rankingTitleTextColor: '#4f6571',
        rankingListBg: '#AADFF0',
        rankingListTextColor: '#4f6571',
        rankingItemBorderColor: '#6cb8cf'
    };

    const darkTheme = {
        background: '#121212',
        textColor: '#ffffff',

        headerBg: '#262626',
        headerTextColor: '#ffffff',

        containerBg: '#121212',

        personalRecordItemTitleTextColor: '#ffffff',
        personalRecordItemContentBg: '#ffffff',
        personalRecordItemContentTextColor: '#121212',

        ratingTextColor: '#ffffff',

        settingBg: '#191919',
        settingItemBg: '#121212',
        settingItemTextColor: '#d7d7d7',
        settingItemBorderColor: '#191919',
        settingItemIconColor: '#d7d7d7',
        settingItemInActiveBg: '#444444',

        versionTextColor: '#d7d7d7',

        modelBg: '#191919',
        modelTextColor: '#d7d7d7',
        modelConfirmBtnBg: '#121212',
        modelConfirmBtnTextColor: '#ffffff',

        functionBtnTextColor: '#ffffff',

        rankingBg: '#191919',
        rankingBorderColor: '#191919',
        rankingTitleTextColor: '#d7d7d7',
        rankingListBg: '#121212',
        rankingListTextColor: '#d7d7d7',
        rankingItemBorderColor: '#191919'
    };

    return darkMode ? darkTheme : lightTheme;
};