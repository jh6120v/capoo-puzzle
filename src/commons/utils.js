import React, { Suspense } from 'react';
import { LazyLoad } from '../styles/common-style';

export const createActionCreator = (namespace) => (actionType) => {
    const type = `${namespace}/${actionType}`;
    const actionCreator = (payload = {}) => ({
        type,
        payload
    });

    actionCreator.type = type;
    Object.freeze(actionCreator);

    return actionCreator;
};

export const waitingRouteComponent = (Component) => () => (
    <Suspense fallback={<LazyLoad>Loading...</LazyLoad>}>
        <Component />
    </Suspense>
);

export const getRandom = (min, max) => Math.floor(Math.random() * (max - min + 1)) + min;

/**
 * 取一逆序列數和為 0 的一維陣列
 * e.g. [1, 2, 3, 4, 5]
 *
 * @param total
 * @returns {{label: number}[]}
 */
export const getTiles = (total) => {
    return Array(total).fill(0).map((value, index) => ({ label: index + 1 }));
};

/**
 * 將數列隨機打散後，並檢查其是否可用(可解)
 *
 * @param tiles
 * @param size
 * @returns {*}
 */
export const getAcceptableTiles = (tiles, size) => {
    let resolvable = false;

    while (!resolvable) {
        tiles.sort(function () {
            return Math.random() > 0.5 ? -1 : 1;
        });

        resolvable = checkResolvable(tiles, size);
    }

    return tiles;
};

/**
 * 檢查數列是否有解，條件 e.g.
 * 1. 拼圖列數為奇數時，逆序列數和應為偶數
 * 2. 拼圖列數為偶數時，逆序列數和的奇偶性與空白磚所在列數的奇偶性守恆
 *
 * @param grids
 * @param size
 * @returns {boolean}
 */
export const checkResolvable = (grids, size) => {
    // 逆序列數和
    let count = 0;

    // 空白磚所在列數
    let spaceX = 0;

    // 找出空白磚列數，並從 grids 中移除
    grids = grids.filter((item, idx) => {
        if (item === grids.length) {
            spaceX = idx % size + 1;
        }

        // 把空白從數列中去除(即數列中最後一數字)
        return item !== grids.length;
    });

    // 計算逆序列數和
    grids.forEach((item, idx, grids) => {
        let j = idx + 1;
        while (j < grids.length) {
            if (item > grids[j]) {
                count++;
            }

            j++;
        }
    });

    return size % 2 ? count % 2 === 0 : count % 2 + spaceX % 2 === 0;
};

/**
 *
 * @param size
 * @returns {{label: number}[]}
 */
export function getInOrderGrids(size) {
    return getTiles(size * size);
}

/**
 *
 * @param size
 * @param originTiles
 * @returns {*}
 */
export const getGrids = (size, originTiles = null) => {
    const tiles = originTiles ? originTiles : getTiles(size * size);

    return getAcceptableTiles(tiles, size);
};
