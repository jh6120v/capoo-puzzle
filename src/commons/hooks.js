import React, { useState, useCallback, useRef, useEffect } from 'react';
import {
    ConfirmButton, ModelContent, ModelFooter, ModelShadow, ModelWrap
} from '../styles/model-style';
import {
    useObservable,
    pluckFirst,
    useObservableState
} from "observable-hooks";
import {
    map,
    filter,
    distinctUntilChanged,
    switchMap,
    repeat,
    withLatestFrom,
    scan,
    tap,
    takeWhile
} from "rxjs/operators";
import { of, animationFrameScheduler } from "rxjs";
import { layoutPosition } from "./utils";

const useModel = (message, confirm) => {
    const [isShown, setShown] = useState(false);
    const showModal = useCallback(() => setShown(true), []);
    const hideModal = useCallback(() => setShown(false), []);

    const ModelBox = () => (
        <div>
            <ModelWrap>
                <ModelContent>{message}</ModelContent>
                <ModelFooter>
                    <ConfirmButton onClick={confirm}>Confirm</ConfirmButton>
                </ModelFooter>
            </ModelWrap>
            <ModelShadow onClick={hideModal} />
        </div>
    );

    return {
        ModelBox,
        isShown,
        showModal,
        hideModal
    };
};

const useTimer = (initialTime = 0, direction = 'forward' , onSuccess) => {
    const [timerState, setTimerState] = React.useState('reset');
    const initialAccumulationValue = direction === 'forward' ? -1 : initialTime + 1;

    const timerState$ = useObservable(pluckFirst, [timerState]);
    const countDown$ = useObservable(() =>
        timerState$.pipe(
            map(state => state === "reset"),
            distinctUntilChanged(),
            switchMap(isReset => {
                console.log("isReset", isReset);

                return isReset
                    ? of(initialTime)
                    : // high accuracy timing
                      // repetitively calculate the diff
                    of(animationFrameScheduler.now(), animationFrameScheduler).pipe(
                        repeat(),
                        // extract seconds
                        map(startTime => ~~((Date.now() - startTime) / 1000)),
                        distinctUntilChanged(),
                        withLatestFrom(timerState$),
                        filter(([, state]) => state === "started"),
                        takeWhile(([val,]) => {
                            // console.log(val, direction, initialTime);
                            return direction === 'forward' || (direction === 'backward' && val <= initialTime);
                        }),
                        tap(([val,]) => {
                            // console.log(val);
                            if (direction === 'backward' && val === initialTime) {
                                onSuccess();
                            }
                        }),
                        // count how many second left
                        scan(timeLeft => {
                            // console.log(timeLeft);
                            return direction === 'backward' ? timeLeft - 1 : timeLeft + 1;
                        }, initialAccumulationValue)
                    );
            })
        )
    );

    const seconds = useObservableState(countDown$, initialTime);

    return {
        timerState,
        setTimerState,
        seconds
    };
};

const usePuzzle = (grids, col) => {
    // 拼圖各方塊絕對定位
    const [layout, setLayout] = useState([]);

    // 拼圖實際寬度(正方形)
    const [puzzleWidth, setPuzzleWidth] = useState(0);

    // 移動次數
    const [move, setMove] = useState(0);

    const puzzleContainer = useRef();
    useEffect(() => {
        // 設定 puzzle 總寬度
        setPuzzleWidth(puzzleContainer.current.clientWidth);

        // 設定拼圖各方塊絕對定位
        setLayout(layoutPosition(puzzleContainer.current.clientWidth, col));
    }, []);

    return {
        puzzleContainer,
        total: col * col,
        puzzleWidth,
        layout,
        moves: [move, setMove]
    }
};

export {
    usePuzzle,
    useModel,
    useTimer
};
