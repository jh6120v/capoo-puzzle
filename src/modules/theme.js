import { createActionCreator } from '../commons/utils';
import { createReducer } from '../store/reducers';

// Actions
const actionCreator = createActionCreator('@@THEME');
const colorModeSet = actionCreator('COLOR_MODE_SET');
const toggleSwitchSet = actionCreator('TOGGLE_SWITCH_SET');

export { colorModeSet, toggleSwitchSet }


// Reducers
const initialState = {
    darkModeEnabled: false,
    colorMode: 'light',
    toggle: () => {}
};

const handlers = {
    [colorModeSet.type]: (state, { payload: { colorMode } }) => ({
        ...state,
        colorMode: colorMode
    }),
    [toggleSwitchSet.type]: (state, { payload: { toggle } }) => ({
        ...state,
        toggle: toggle
    })
};

// create and export as default
const reducers = createReducer(initialState, handlers);
export default reducers;
