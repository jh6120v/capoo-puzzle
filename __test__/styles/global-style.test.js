import React from 'react';
import { render } from '@testing-library/react';
import GlobalStyle from '../../src/styles/global-style';
import 'jest-styled-components';
import { ThemeProvider } from 'styled-components';
import { theme } from '../../src/commons/theme';

describe('test global style.', () => {
    it('global style snap shot', () => {
        const { container } = render(
            <ThemeProvider theme={theme(false)}>
                <GlobalStyle />
            </ThemeProvider>
        );

        expect(container.textContent).toMatchSnapshot();
    });
});
