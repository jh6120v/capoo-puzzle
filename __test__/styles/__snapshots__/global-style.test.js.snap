// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test global style. global style snap shot 1`] = `""`;
